File: examples/gemini/tracing.py
"""
This is an example for leveraging MLflow's auto tracing capabilities for Gemini.

For more information about MLflow Tracing, see: https://mlflow.org/docs/latest/llms/tracing/index.html
"""

import os

import mlflow

# Turn on auto tracing for Gemini by calling mlflow.gemini.autolog()
mlflow.gemini.autolog()

# Import the SDK and configure your API key.
from google import genai

client = genai.Client(api_key=os.environ["GEMINI_API_KEY"])

# Use the generate_content method to generate responses to your prompts.
response = client.models.generate_content(
    model="gemini-1.5-flash", contents="The opposite of hot is"
)
print(response.text)

# Also leverage the chat feature to conduct multi-turn interactions
chat = client.chats.create(model="gemini-1.5-flash")
response = chat.send_message("In one sentence, explain how a computer works to a young child.")
print(response.text)
response = chat.send_message("Okay, how about a more detailed explanation to a high schooler?")
print(response.text)

# Count tokens for your statement
response = client.models.count_tokens("The quick brown fox jumps over the lazy dog.")
print(response.total_tokens)

# Generate text embeddings for your content
text = "Hello world"
result = client.models.embed_content(model="text-embedding-004", contents=text)
print(result["embedding"])

File: libs/typescript/integrations/gemini/example/tracing.ts

import * as mlflow from 'mlflow-tracing';
import { GoogleGenAI } from '@google/genai';
import { tracedGemini } from 'mlflow-gemini';

async function main() {
  mlflow.init({
    trackingUri: 'http://localhost:5000',
    experimentId: '252237888791228129'
  });

  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error('GEMINI_API_KEY environment variable is not set.');
  }

  const client = tracedGemini(new GoogleGenAI({ apiKey }));

  const response = await client.models.generateContent({
    model: 'gemini-2.0-flash-001',
    contents: 'Why is the sky blue?'
  });

  console.log(response.text);

//   await mlflow.flushTraces();
}

main().catch((error) => {
  console.error('Error running Gemini tracing example:', error);
  process.exitCode = 1;
});